source/common/common/trie.h

```diff
@@ -0,0 +1,111 @@
+   */
+  void emplace(const Key& key, Value value);
+
+  /**
+   * Lookup a value by key.
+   * @param key the path name to look up
+   * @return a std::pair consisting of the value if found and a boolean
+   *         indicating if the match was exact.
+   */
+  std::pair<Value, bool> find(const Key& key) const;
+
+private:
+  std::pair<Value, bool> find(std::vector<Key>& path_components) const;
+  void emplace(std::vector<Key>& path_components, Value value);
+  Value value() const { return value_; }
+  void set_value(Value value) { value_ = value; }
+
+  // By holding pointers to the TriedNodes we don't have to worry that they're
+  // incomplete types.
+  std::unordered_map<Key, std::unique_ptr<TrieNode<Key, Value, Tokenizer>>> children_;
+  Value value_;
+  Tokenizer tokenizer_;
+};
+
+template <class Key, class Value, class Tokenizer>
+void TrieNode<Key, Value, Tokenizer>::emplace(const Key& key, Value value) {
+  std::vector<Key> path_components = tokenizer_.tokenize(key);
+  // Keys come in big endian but we reverse them for vector iteration
+  // efficiency.
+  std::reverse(path_components.begin(), path_components.end());
+  emplace(path_components, value);
+}
+
+template <class Key, class Value, class Tokenizer>
+void TrieNode<Key, Value, Tokenizer>::emplace(std::vector<Key>& path_components, Value value) {
+  if (path_components.empty()) {
+    return;
+  }
+  Key name(path_components.back());
+  path_components.pop_back();
+  if (children_.find(name) == children_.end()) {
+    std::unique_ptr<TrieNode<Key, Value, Tokenizer>> node(
+        new TrieNode<Key, Value, Tokenizer>(nullptr));
+    children_[name] = std::move(node);
+  }
+  TrieNode<Key, Value, Tokenizer>* node = children_[name].get();
+  if (path_components.empty()) {
+    node->set_value(value);
+  } else {
+    node->emplace(path_components, value);
+  }
+}
+
+template <class Key, class Value, class Tokenizer>
+std::pair<Value, bool> TrieNode<Key, Value, Tokenizer>::find(const Key& key) const {
+  std::vector<Key> path_components = tokenizer_.tokenize(key);
+  // Keys come in big endian but we reverse them for vector iteration
+  // efficiency.
+  std::reverse(path_components.begin(), path_components.end());
+  return find(path_components);
+}
+
+template <class Key, class Value, class Tokenizer>
+std::pair<Value, bool>
+TrieNode<Key, Value, Tokenizer>::find(std::vector<Key>& path_components) const {
+  if (path_components.empty()) {
+    // An exact match requires that we have a non-false value at the very end of
+    // our path traversal.
+    return std::make_pair(value_, value_ ? true : false);
+  }
+  bool exact_match = false;
+  Value value = value_;
+  Key component(path_components.back());
+  path_components.pop_back();
+  if (children_.find(component) != children_.end()) {
+    const TrieNode<Key, Value, Tokenizer>* node = children_.find(component)->second.get();
+    std::pair<Value, bool> retval = node->find(path_components);
```

Comment:
Optimally, this function would be written without recursion, in case we have deep nesting. I don't think it matters that much but I would drop a comment in.